crypto/thread/arch.c

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2019-2021 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/configuration.h>
+#include <openssl/crypto.h>
+
+#if defined(OPENSSL_THREADS)
+# include <internal/thread_arch.h>
+
+CRYPTO_THREAD * crypto_thread_native_start(CRYPTO_THREAD_ROUTINE_CB routine,
+                                           void *data, int joinable)
+{
+    struct crypto_thread_st *handle;
+
+    if (routine == NULL)
+        return NULL;
+
+    handle = OPENSSL_zalloc(sizeof(*handle));
+    if (handle == NULL)
+        return NULL;
+
+    handle->lock = CRYPTO_MUTEX_create();
```

Comment:
Would rwlock be better?

Reply:
rwlock discussed in this comment: https://github.com/openssl/openssl/pull/12255/files#r922710712